sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -591,8 +587,8 @@ object DateTimeUtils {
     }
     // using milliseconds can cause precision loss with more than 8 digits
     // we follow Hive's implementation which uses seconds
-    val secondsInDay1 = MILLISECONDS.toSeconds(millis1 - daysToMillis(date1, zoneId))
```

Comment:
shall we call `Math.floorDiv`?

Reply:
Highly likely, yes. I will prepare a separate fix.

Reply:
The error of rounding is invisible in dividing by `DAYS.toSeconds(31)`. At least, I haven't reproduce the issue yet.

Reply:
ah, this is "seconds in day", so it's always positive.